p2p/transport/webrtc/stream_read.go

```diff
@@ -0,0 +1,166 @@
+package libp2pwebrtc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+	"github.com/libp2p/go-msgio/pbio"
+	"github.com/pion/datachannel"
+)
+
+type (
+	webRTCStreamReader struct {
+		stream *webRTCStream
+		state  *webRTCStreamReaderState
+
+		deadline    time.Time
+		deadlineMux sync.Mutex
+
+		closeOnce sync.Once
+	}
+
+	webRTCStreamReaderState struct {
+		sync.Mutex
+
+		Reader pbio.Reader
+		Buffer []byte
+
+		stream *webRTCStream
+	}
+)
+
+// Read from the underlying datachannel. This also
+// process sctp control messages such as DCEP, which is
+// handled internally by pion, and stream closure which
+// is signaled by `Read` on the datachannel returning
+// io.EOF.
+func (r *webRTCStreamReader) Read(b []byte) (int, error) {
+	if len(b) == 0 {
+		return 0, nil
+	}
+
+	var (
+		readErr error
+		read    int
+	)
+	for read == 0 && readErr == nil {
+		if r.stream.isClosed() {
+			return 0, io.ErrClosedPipe
+		}
+
+		readDeadline, hasReadDeadline := r.getReadDeadline()
+		if hasReadDeadline {
+			// check if deadline exceeded
+			if readDeadline.Before(time.Now()) {
+				if err := r.stream.getCloseErr(); err != nil {
+					log.Debugf("[1] deadline exceeded: closeErr: %v", err)
+				} else {
+					log.Debug("[1] deadline exceeded: no closeErr")
+				}
+				return 0, os.ErrDeadlineExceeded
+			}
+		}
+
+		read, readErr = r.state.readMessage(b)
+	}
+
+	return read, readErr
+}
+
+func (r *webRTCStreamReaderState) readMessage(b []byte) (int, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	read := copy(b, r.Buffer)
+	r.Buffer = r.Buffer[read:]
+	remaining := len(r.Buffer)
+
+	if remaining == 0 && !r.stream.stateHandler.AllowRead() {
+		if closeErr := r.stream.getCloseErr(); closeErr != nil {
+			log.Debugf("[2] stream closed: %v", closeErr)
+			return read, closeErr
+		}
+		log.Debug("[2] stream empty")
+		return read, io.EOF
+	}
+
+	if read > 0 {
+		return read, nil
+	}
+
+	// read from datachannel
+	var msg pb.Message
+	err := r.Reader.ReadMsg(&msg)
+	if err != nil {
+		// This case occurs when the remote node goes away
+		// without writing a FIN message
+		if errors.Is(err, io.EOF) {
+			r.stream.Reset()
+			return read, io.ErrClosedPipe
+		}
+		if errors.Is(err, os.ErrDeadlineExceeded) {
+			// if the stream has been force closed or force reset
+			// using SetReadDeadline, we check if closeErr was set.
```

Comment:
Deadlines don't close / reset streams.

Reply:
They don't. I've worded the  comment poorly. Please take a look at the following snippet. https://github.com/libp2p/go-libp2p/pull/1999/files/79193f8bd8573796570146a1d5208c9094bbc4c7#diff-66b81e98e3f9cb27698e7caffd2f14850b8eea6a709a4cab356592f9f8283817R536-R545

If we call `Close` on a detached Pion datachannel when a `Read` call is in progress, it will not cause the read call to return immediately. The `Read` only returns an `io.EOF` when the remote closes the stream in response to the local close that is sent.

Consider the following flow:
1. Let A, B be two peers one opposite sides of the stream.
2. A initiates a `Read` and is blocked until B sends data or closes the stream.
3. In a separate routine, A decides to `Close` the stream.
3. Consider the following alternatives:
a. Happy path: B is also reading on the stream. It receives the `Close` message sent by A, and responds with a `Close` message. This causes the `Read` initiated in step 2 to return gracefully with an EOF.
b. If B is not reading on the stream, it will not read the close message sent by A. Therefore it does not respond with a close message, causing the Read initiated in step 2 to block indefinitely.

To mitigate this issue, we set a `closeErr` variable on the stream and  set a ReadDeadline to a past time. This causes any reads in progress to return immediately. For this, if a Read exits with ErrDeadlineExceeded, we check if it was caused by `closeErr` being set.